### PR TITLE
Feature/save servers

### LINKED ISSUE #354: Saved server list
It would be very useful to have a list of bookmarks for multiple servers with saved URLs and credentials. Currently, the user has to enter the url and login everytime they switch servers.

### PR DESCRIPTION
**Pull request by jimdogx**
Saves successfully connected to servers.  Next time a user selects "Change Server" from the option menu, simply clicking on the server entry will log them right in (assuming they have logged in at least once before).

**Changes**
After the client scans for local servers, any saved servers will be added to the list as well.

Other changes, "options" are now enabled on the server list screen.  The option button will only work if the user is currently selecting a "Saved" server (not a locally discovered one).  If on a "Saved" entry, clicking on the options will allow the user to delete the saved entry.

Issues
Fixes #354  

Notes: I looked in to using a token, but the client is already saving the user's password.  If / when password saving is refactored in favor of a token, this new code an easily be changed from "password" to "token".

### CODE REVIEW

**Comment by crobibero** [comment]
(I didn't mean to hit update branch, sorry)

**Comment by neilsb** [comment]
Had a quick play with it, and looking good.

Would it be possible to store the server name when saving the details?  At the moment if you get multiple saved servers then there are multiple in the list with the name "_Saved_".    Might be nice to show the name as "_My Server (Saved)_"  or something?

The saving of the username and password only appears to work if you go back into the same saved server.  i.e.  If I have 2 saved servers and am logged into Server 1, and then choose _Change Server_, I can log back into Server 1 fine, but selecting Server 2 requires the username and password again.    Then I can change server and go back into server 2  fine, but changing server and  selecting Server 1 then requires password, etc.

This was just a quick play though, so maybe I messed up somewhere.  I'll play with it some more in the next few days.

**Comment by jimdogx** [comment]
Thanks for testing!  I'll look in to getting the server name saved (in theory that should be easy).

I never tested with more than one "saved" server.  Only with a locally discovered server and one "saved" server, so you may have found an issue with more than one "saved" server.  I'll setup another Jellyfin server this weekend and play around with that scenario 👍

File: components/config/SetServerScreen.brs [hunk 1]
```diff
@@ -63,6 +68,25 @@ sub ScanForServersComplete(event)
...
     end for
+
+    'load any previously logged in to servers as well (if they aren't already discovered on the local network)
+    saved = get_setting("saved_servers")
+    if saved <> invalid
+        savedServers = ParseJson(saved)
+        for each server in savedServers.serverList
+            alreadyListed = false
+            for each listed in m.servers
+                if listed.baseUrl = server.baseUrl
```
Comment (crobibero) [inline components/config/SetServerScreen.brs]: Does this account for users changing the url case?

Comment (jimdogx) [reply components/config/SetServerScreen.brs]: No it does not.  Why oh why would a user do that!?!? 😄    I'll add some defensive code.

**Comment by jimdogx** [comment]
@neilsb Updated and tested.  Thanks again for testing it out for me 👍 

If I ever hear someone complain about having to enter URLs again...  I had to enter / delete / re-enter URLs for testing about 100 times!!! 😄

File: components/config/SetServerScreen.brs [hunk 2]
```diff
@@ -38,6 +37,13 @@ function onKeyEvent(key as string, press as boolean) as boolean
         'focus the submit button from serverUrl
     else if key = "down" and m.serverUrlContainer.hasFocus()
         m.submit.setFocus(true)
+    else if key = "options"
+        serverName = m.serverPicker.content.getChild(m.serverPicker.itemFocused).name
+        if m.servers.Count() > 0 and Instr(1, serverName, "Saved") > 0
```
Comment (crobibero) [inline components/config/SetServerScreen.brs:42]: What happens if a server is called "Saved"?

File: source/Main.brs [hunk 3]
```diff
@@ -406,6 +408,50 @@ function LoginFlow(startOver = false as boolean)
...
+    saved = get_setting("saved_servers")
+    entryCount = 0
+    addNewEntry = true
+    savedServers = { serverList: [] }
+    if saved <> invalid 
+        savedServers = ParseJson(saved)
+        entryCount = savedServers.serverList.Count() 
+        if savedServers.serverList <> invalid and entryCount > 0
+            for each item in savedServers.serverList
+                if item.baseUrl = LCase(server) 'Saved server data is always lowercase
```
Comment (crobibero) [inline source/Main.brs]: Can the `LCase(server)` be done when getting the server from settings to avoid manipulating the string multiple times?

File: components/config/SetServerScreen.brs (hunk 2 above)
Comment (jimdogx) [reply components/config/SetServerScreen.brs:42]: Then they will have the option to delete it (optionally).  And if they do, they will have to re-enter their credentials (no different / worst than before this update).

File: source/Main.brs (hunk 3 above)
Comment (jimdogx) [reply source/Main.brs]: Yes.

File: source/Main.brs [hunk 4]
```diff
@@ -406,6 +408,56 @@ function LoginFlow(startOver = false as boolean)
...
+                    addNewEntry = false
+                    exit for
+                end if
+            end for
+        end if
+    end if
+
+    if addNewEntry
+        if entryCount = 0
+            set_setting("saved_servers", FormatJson({ serverList: [{ name: m.serverSelection, baseUrl: LCase(server), iconUrl: "pkg:/images/logo-icon120.jpg", iconWidth: 120, iconHeight: 120 }] }))
```
Comment (crobibero) [inline source/Main.brs]:
```suggestion
            set_setting("saved_servers", FormatJson({ serverList: [{ name: m.serverSelection, baseUrl: server, iconUrl: "pkg:/images/logo-icon120.jpg", iconWidth: 120, iconHeight: 120 }] }))
```

File: source/Main.brs [hunk 5]
```diff
@@ -406,6 +408,56 @@ function LoginFlow(startOver = false as boolean)
...
+        else
+            savedServers.serverList.Push({ name: m.serverSelection, baseUrl: LCase(server), iconUrl: "pkg:/images/logo-icon120.jpg", iconWidth: 120, iconHeight: 120 })
```
Comment (crobibero) [inline source/Main.brs]:
```suggestion
            savedServers.serverList.Push({ name: m.serverSelection, baseUrl: server, iconUrl: "pkg:/images/logo-icon120.jpg", iconWidth: 120, iconHeight: 120 })
```

File: source/Main.brs (hunk 4 above)
Comment (jimdogx) [reply source/Main.brs]: Too excited getting this stuff working that I miss the easy stuff.

**Comment by neilsb** [comment]
Had another testing session, and it's working really well!  On thing I did pick up is that the "Sign Out" option from the menu will immediately sign you back in again (when using hidden users - the default).

So the only way to log in as another user on a server is to select change server, delete the server and re-enter it.  And if it's a server detected on the network there's no way of changing the signed in user as it can' be deleted short of uninstalling and reinstalling the app.

I wonder if we should have a checkbox under the username/password entry for "Save login" or something like that, (which can default to being checked?) but would allow someone not to save the login if they desired.    The server could always be saved but they could choose whether the login should be saved.

And the "Sign Out" option could remove the saved username and password from a server, forcing another login the next time it is selected?

This is really for discussion;  just trying to think of use cases.   Can see people in a single household always wanting to use separate accounts, whether to track viewing progress or restrict some users access to particular libraries.

**Comment by jimdogx** [comment]
> Had another testing session, and it's working really well! On thing I did pick up is that the "Sign Out" option from the menu will immediately sign you back in again (when using hidden users - the default)...

Thanks again!

> ...And the "Sign Out" option could remove the saved username and password from a server, forcing another login the next time it is selected?

I like this option.  I'll start there and see how the user experience is before investigating the checkbox option.

I bought a new Roku just for development / testing (only have one device in the house right now that I'm timesharing) so hopefully I'll get better at testing going forward 😄

**Comment by jimdogx** [comment]
Do I win the award for longest Jellyfin-Roku PR ever?

Updates:
* New Checkbox for saving credentials.  I removed an unused "Label" in ConfigScene at the same time.
* Changing servers does not remove any credentials, but Signing Out does.
* ClearScenes when hitting back from Sign in page after Signing out (this is currently an issue on Master as well)

Known issues:
I noticed the following sequence causes a crash:
_Sign out -> back -> select same server -> sign in_
However, this also happens on Master, so I don't think this is related to anything I've added and needs to be looked in to separately.

**Comment by jimdogx** [comment]
Also wondering if we need better wording for "Save Credentials?" As this setting / checkbox is only related to switching between servers.

It might be confusing if a user unchecks the box, logs in and then the next day sees that they are still re-logged in when they launch the app.

**Comment by bisby** [comment]
Definitely don't want to throw a wrench in anything that is happening here. But I do want to point out that this was (poorly) implemented long ago, and then removed: https://github.com/jellyfin/jellyfin-roku/commit/29d65928f5137de8943d089bde8e9f0f8b805dc1#diff-9aea1db3489117f1a420a0ee1f691d71afd772969ea9caa63c1b8477ab7087ff here's the commit that added it all.

The biggest issue (at the time at least) was that Jellyfin auth uses the device ID as part of the auth request. And when an auth request comes in with device id = X... all other auth tokens for device X are invalidated. So if you signed in with ProfileA. and then went to add ProfileB, ProfileA would be signed out. Full stop. No room for discussion. "Just implement the re-log-in flow using PINs" was what I was told. I don't use a PIN, and my password is 32 characters with random symbols and is a nightmare to input. So instead, I made it so the device ID is random: https://github.com/jellyfin/jellyfin-roku/blob/master/source/api/baserequest.brs#L146 . This probably causes all sorts of other issues related to device tracking for the server, but it means that each profile has a separate device ID, so you can have 2 profiles logged in on the same server at a time. This might no longer be a necessity.

tl;dr - this was partially done before in a quick and dirty way, and there might be some bad legacy stuff hanging out. Watch out for speed bumps.

**Comment by neilsb** [comment]
That's interesting and the deviceId thing explains some things about the server not having saved profiles for the Roku client even when they were being sent up 😄 

I don't think any of this should affect this PR, since it's just saving the server and (at most) a single login for that server. As far as I can remember, there's no functionality to have 2 users details stored on the Jellyfin client at the moment.  

Looking at having a non-changing deviceId should probably be looked into at some point though.

**Comment by bisby** [comment]
The first PR definitely added multiple saved profiles. I dont know when (or why) it was removed. Glad to see it making a comeback though.

**Review by neilsb — APPROVED** [review]
Seem to work just fine now.  All looking great!

**Comment by neilsb** [comment]
> It might be confusing if a user unchecks the box, logs in and then the next day sees that they are still re-logged in when they launch the app.

I see what you mean, but not sure it's something that needs addresses right now.  As part of updating the general app design, the login pages will be looked at, and I think it would be fine to leave it till then to consider how it could be changed.  We can see what the user feedback is.
